protocol/auditlog_test.go

```diff
@@ -0,0 +1,247 @@
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+}
+
+func TestUpdateHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// update the directory so we can update the audit log
+	d.Update()
+	err = aud.Update("test-server", d.LatestSTR())
+
+	if err != nil {
+		t.Fatal("Error updating the server history")
+	}
+}
+
+func TestInsertPriorHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+
+	// create 10 epochs
+	priorSTRs := make(map[uint64]*DirSTR)
+	for i := 0; i < 10; i++ {
+		priorSTRs[d.LatestSTR().Epoch] = d.LatestSTR()
+		d.Update()
+	}
+
+	// now insert
+	err := aud.Insert("test-server", pk, priorSTRs, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history with prior STRs")
+	}
+}
+
+func TestInsertExistingHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// let's make sure that we can't re-insert a new server
+	// history into our log
+	err = aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when inserting an existing server history")
+	}
+}
+
+func TestUpdateUnknownHistory(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	// let's make sure that we can't re-insert a new server
+	// history into our log
+	err = aud.Update("unknown", d.LatestSTR())
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when updating an unknown server history")
+	}
+}
+
+func TestGetLatestObservedSTR(t *testing.T) {
+	// let's just create basic test directory and an empty audit log
+	d, pk := NewTestDirectory(t, true)
+	aud := NewAuditLog()
+	err := aud.Insert("test-server", pk, nil, d.LatestSTR())
+	if err != nil {
+		t.Fatal("Error inserting new server history")
+	}
+
+	res, err := aud.GetObservedSTRs(&AuditingRequest{
+		DirectoryAddr: "test-server",
+		StartEpoch:    uint64(d.LatestSTR().Epoch),
+		EndEpoch:      uint64(d.LatestSTR().Epoch)})
+	if err != ReqSuccess {
+		t.Fatal("Unable to get latest observed STR")
+	}
+
+	obs := res.DirectoryResponse.(*STRHistoryRange)
+	if len(obs.STR) != 1 {
+		t.Fatal("Expect returned STR to be not nil")
+	}
+	if obs.STR[0].Epoch != d.LatestSTR().Epoch {
+		t.Fatal("Unexpected epoch for returned STR")
+	}
```

Comment:
Please add a testcase for https://github.com/coniks-sys/coniks-go/pull/155/files/f63dbb0b265021712fbdcbc78b8d9f84aee6b52f#r124956103.